libs/vkd3d/command.c

```diff
@@ -288,164 +288,101 @@ static HRESULT vkd3d_enqueue_timeline_semaphore(struct vkd3d_fence_worker *worke
+    if (FAILED(hr = d3d12_fence_signal(fence->waiting_fence.fence, counter_value)))
+        ERR("Failed to signal D3D12 fence, hr %#x.\n", hr);
 
-        if (i != j)
-        {
-            worker->vk_semaphores[j] = worker->vk_semaphores[i];
-            worker->semaphore_wait_values[j] = worker->semaphore_wait_values[i];
-            worker->fences[j] = worker->fences[i];
-        }
-        ++j;
-    }
-    worker->fence_count = j;
+    d3d12_fence_dec_ref(fence->waiting_fence.fence);
 }
 
 static void *vkd3d_fence_worker_main(void *arg)
 {
     struct vkd3d_fence_worker *worker = arg;
+    struct vkd3d_enqueued_fence *fences;
+    uint32_t i, fence_count;
+    bool do_exit;
     int rc;
 
     vkd3d_set_thread_name("vkd3d_fence");
 
     for (;;)
     {
-        vkd3d_wait_for_gpu_timeline_semaphores(worker);
+        if ((rc = pthread_mutex_lock(&worker->mutex)))
+        {
+            ERR("Failed to lock mutex, error %d.\n", rc);
+            break;
+        }
 
-        if (!worker->fence_count || vkd3d_atomic_uint32_load_explicit(&worker->enqueued_fence_count, vkd3d_memory_order_acquire))
+        if (!worker->enqueued_fence_count)
         {
-            if ((rc = pthread_mutex_lock(&worker->mutex)))
+            if ((rc = pthread_cond_wait(&worker->cond, &worker->mutex)))
             {
-                ERR("Failed to lock mutex, error %d.\n", rc);
+                ERR("Failed to wait on condition variable, error %d.\n", rc);
+                pthread_mutex_unlock(&worker->mutex);
                 break;
             }
+        }
 
-            if (worker->pending_fence_destruction)
-            {
-                pthread_cond_broadcast(&worker->fence_destruction_cond);
-                worker->pending_fence_destruction = false;
-            }
+        fence_count = worker->enqueued_fence_count;
+        fences = worker->enqueued_fences;
+        do_exit = worker->should_exit;
 
-            if (worker->enqueued_fence_count)
-            {
-                vkd3d_fence_worker_move_enqueued_fences_locked(worker);
-            }
-            else
-            {
-                if (worker->should_exit)
-                {
-                    pthread_mutex_unlock(&worker->mutex);
-                    break;
-                }
+        worker->enqueued_fence_count = 0;
+        worker->enqueued_fences = NULL;
+        worker->enqueued_fences_size = 0;
 
-                if ((rc = pthread_cond_wait(&worker->cond, &worker->mutex)))
-                {
-                    ERR("Failed to wait on condition variable, error %d.\n", rc);
-                    pthread_mutex_unlock(&worker->mutex);
-                    break;
-                }
-            }
+        pthread_mutex_unlock(&worker->mutex);
 
-            pthread_mutex_unlock(&worker->mutex);
-        }
+        for (i = 0; i < fence_count; i++)
+            vkd3d_wait_for_gpu_timeline_semaphore(worker, &fences[i]);
+
+        vkd3d_free(fences);
```

Comment:
We should probably just pop one fence per iteration to avoid forcing allocation on every single wait.

Reply:
Went with the swap thing.